### ISSUE TITLE
Socket initialization in BIO_sock_init isn't thread-safe on Windows

### ISSUE THREAD

**Issue by bplu4t2f** [openssl/openssl#16727]
[See also:  BIO_sock_init() needs a concurrency lock on Win32. #7346](https://github.com/openssl/openssl/issues/7346)

A way to repro is starting two separate threads on program startup, both of which do BIO_accept on an accept bio.

One of them will almost certainly fail with `Either the application has not called WSAStartup, or WSAStartup failed.`.

Adding `BIO_sock_init` before starting the threads fixes the issue. For the time being, this should at least be documented or otherwise mentioned as a *really important* initialization call on Windows (e.g. in the wiki).

Opening new issue because the linked issue was closed by the author even though it's not fixed.

**Comment by ChrisDenton** [comment]
It seems like the problem is just that `wsa_init_done` is set *before* the call to `WSAStartup` is complete.

https://github.com/openssl/openssl/blob/005f545f6451a16b44aa374e53202f39114cf8d5/crypto/bio/bio_sock.c#L142-L151

Moving it after would fix the immediate issue at the cost of potentially calling `WSAStartup` multiple times if threads race with each other. Some synchronisation could be used to ensure either `WSAStartup` is only called once or else `WSACleanup` is called if another thread initialised first (the latter case can probably be done with just an atomic).

**Comment by nhorman** [comment]
how are you creating the sockets?  I ask because creating the sockets vi BIO_socket already implicitly calls BIO_sock_init for you (implying a call to WSAStartup).  If you are creating the sockets using Windows API call and assigning them to a BIO with BIO_set_fd implies that the application is taking responsibility for any platform specific initialization.

That should certainly be documented, but it seems like the Windows API documentation is the appropriate place for that:
https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-socket

**Comment by bplu4t2f** [comment]
@nhorman The bug isn't that `BIO_sock_init` isn't being called. It's that `BIO_sock_init` may be called multiple times simultaneously as a result of calling some other BIO functions (for example accept or do_connect) and the `BIO_sock_init` function may fail because it doesn't handle that scenario correctly.

**Comment by nhorman** [comment]
hmm, I see.  Yeah, we should probably move the windows bits of BIO_sock_init

**Comment by nhorman** [comment]
Give this change a try please.

```
diff --git a/crypto/bio/bio_sock.c b/crypto/bio/bio_sock.c
index be1a74620f..8725a739bd 100644
--- a/crypto/bio/bio_sock.c
+++ b/crypto/bio/bio_sock.c
@@ -134,68 +134,77 @@ LPFN_WSARECVMSG bio_WSARecvMsg;
 LPFN_WSASENDMSG bio_WSASendMsg;
 # endif
 
-int BIO_sock_init(void)
-{
 # ifdef OPENSSL_SYS_WINDOWS
+static CRYPTO_ONCE bio_sock_init = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_bio_sock_init)
+{
     static struct WSAData wsa_state;
 
-    if (!wsa_init_done) {
-        wsa_init_done = 1;
-        memset(&wsa_state, 0, sizeof(wsa_state));
-        /*
-         * Not making wsa_state available to the rest of the code is formally
-         * wrong. But the structures we use are [believed to be] invariable
-         * among Winsock DLLs, while API availability is [expected to be]
-         * probed at run-time with DSO_global_lookup.
-         */
-        if (WSAStartup(0x0202, &wsa_state) != 0) {
-            ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
-                           "calling wsastartup()");
-            ERR_raise(ERR_LIB_BIO, BIO_R_WSASTARTUP);
-            return -1;
-        }
+    wsa_init_done = 1;
+    memset(&wsa_state, 0, sizeof(wsa_state));
+    /*
+     * Not making wsa_state available to the rest of the code is formally
+     * wrong. But the structures we use are [believed to be] invariable
+     * among Winsock DLLs, while API availability is [expected to be]
+     * probed at run-time with DSO_global_lookup.
+     */
+    if (WSAStartup(0x0202, &wsa_state) != 0) {
+        ERR_raise_data(ERR_LIB_SYS, get_last_socket_error(),
+                       "calling wsastartup()");
+        ERR_raise(ERR_LIB_BIO, BIO_R_WSASTARTUP);
+        return 0;
+    }
 
-        /*
-         * On Windows, some socket functions are not exposed as a prototype.
-         * Instead, their function pointers must be loaded via this elaborate
-         * process...
-         */
+    /*
+     * On Windows, some socket functions are not exposed as a prototype.
+     * Instead, their function pointers must be loaded via this elaborate
+     * process...
+     */
 #  ifdef BIO_HAVE_WSAMSG
-        {
-            GUID id_WSARecvMsg = WSAID_WSARECVMSG;
-            GUID id_WSASendMsg = WSAID_WSASENDMSG;
-            DWORD len_out = 0;
-            SOCKET s;
-
-            s = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
-            if (s != INVALID_SOCKET) {
-                if (WSAIoctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER,
-                             &id_WSARecvMsg, sizeof(id_WSARecvMsg),
-                             &bio_WSARecvMsg, sizeof(bio_WSARecvMsg),
-                             &len_out, NULL, NULL) != 0
-                    || len_out != sizeof(bio_WSARecvMsg))
-                    bio_WSARecvMsg = NULL;
-
-                if (WSAIoctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER,
-                             &id_WSASendMsg, sizeof(id_WSASendMsg),
-                             &bio_WSASendMsg, sizeof(bio_WSASendMsg),
-                             &len_out, NULL, NULL) != 0
-                    || len_out != sizeof(bio_WSASendMsg))
-                    bio_WSASendMsg = NULL;
-
-                closesocket(s);
-            }
+    {
+        GUID id_WSARecvMsg = WSAID_WSARECVMSG;
+        GUID id_WSASendMsg = WSAID_WSASENDMSG;
+        DWORD len_out = 0;
+        SOCKET s;
+
+        s = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+        if (s != INVALID_SOCKET) {
+            if (WSAIoctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER,
+                         &id_WSARecvMsg, sizeof(id_WSARecvMsg),
+                         &bio_WSARecvMsg, sizeof(bio_WSARecvMsg),
+                         &len_out, NULL, NULL) != 0
+                || len_out != sizeof(bio_WSARecvMsg))
+                bio_WSARecvMsg = NULL;
+
+            if (WSAIoctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER,
+                         &id_WSASendMsg, sizeof(id_WSASendMsg),
+                         &bio_WSASendMsg, sizeof(bio_WSASendMsg),
+                         &len_out, NULL, NULL) != 0
+                || len_out != sizeof(bio_WSASendMsg))
+                bio_WSASendMsg = NULL;
+
+            closesocket(s);
         }
+    }
 #  endif
+    return 1;
+}
+#endif
+
+int BIO_sock_init(void)
+{
+# ifdef OPENSSL_SYS_WINDOWS
+    if (!RUN_ONCE(&bio_sock_init, do_bio_sock_init)) {
+            return -1;
     }
-# endif                         /* OPENSSL_SYS_WINDOWS */
+#endif
+
 # ifdef WATT32
     extern int _watt_do_exit;
     _watt_do_exit = 0;          /* don't make sock_init() call exit() */
     if (sock_init())
         return -1;
 # endif
-
     return 1;
 }
 
```

**Comment by davidben** [comment]
Even if OpenSSL's use is internally synchronized, OpenSSL is a library in a broader application that might itself call WSAStartup unsynchronized with OpenSSL. So either way, you probably should document what parts of OpenSSL might trigger WSAStartup so the application knows to synchronize its initialization.
